### ISSUE TITLE
Add SELU (Scaled ELU)

### ISSUE THREAD

**Issue by edubart** [torch/nn#1240]
Add scaled exponential linear unit (SELU) activation function as intruduced in
[Self-Normalizing Neural Networks](https://arxiv.org/abs/1706.02515v1)

pytorch already added it, see
https://github.com/pytorch/pytorch/pull/1769/files

alpha dropout would be good to have too, pytorch added it too, see
https://github.com/pytorch/pytorch/pull/1775/files

**Comment by kyuusaku** [comment]
`
require 'nn'
def SELU(inplace):
    alpha = 1.6732632423543772848170429916717
    scale = 1.0507009873554804934193349852946
    return nn.Sequential()
                    :add(nn.ELU(alpha, inplace))
                    :add(nn.MulConstant(scale, true))
`

**Comment by kyuusaku** [comment]
I try to write the AlphaDropout module.

```lua
local AlphaDropout, Parent = torch.class('nn.AlphaDropout', 'nn.Module')

--[[
During training, Dropout masks parts of the input using binary samples 
from a bernoulli distribution. 
Each input element has a probability of p of being dropped.
]]

function AlphaDropout:__init(p)
   Parent.__init(self)
   self.p = p or 0.5
   if self.p >= 1 or self.p < 0 then
      error('<Dropout> illegal percentage, must be 0 <= p < 1')
   end
   self.train = true
   self.alpha = -1.7580993408473766
   self.keep_prob = 1 - self.p
   self.a = (self.keep_prob + self.alpha^2 * self.keep_prob *(1 - self.keep_prob))^(-0.5)
   self.b = -self.a * self.alpha * (1 - self.keep_prob)
   self.noise = torch.Tensor()
end

function AlphaDropout:updateOutput(input)
   self.output:resizeAs(input):copy(input)
   if self.p > 0 then
      if self.train then
         self.noise:resizeAs(input)
         self.noise:bernoulli(self.keep_prob)
         self.output:maskedFill(1-self.noise, self.alpha)
         self.output:mul(a):add(b)
      end
   end
   return self.output
end

function AlphaDropout:updateGradInput(input, gradOutput)
   self.gradInput:resizeAs(gradOutput):copy(gradOutput)
   if self.train then
      if self.p > 0 then
         self.gradInput:cmul(self.noise):mul(self.a)
      end
   end
   return self.gradInput
end

function AlphaDropout:setp(p)
   self.p = p
end

function AlphaDropout:__tostring__()
   return string.format('%s(%f)', torch.type(self), self.p)
end


function AlphaDropout:clearState()
   if self.noise then
      self.noise:set()
   end
   return Parent.clearState(self)
end
```

**Comment by kyuusaku** [comment]
Here is the test code for AlphaDropout. However, I do not know how to test the back propagation. Can anyone give some tips? Many thanks.

```lua
-- test_AlphaDropout
require 'nn'
require './AlphaDropout'
-- generate random tensor with zero mean and unit std

local mytest = torch.TestSuite()
local tester = torch.Tester()

local precision = 1e-5

function mytest.AlphaDropout_2()
   local input = torch.randn(5000)
   local mean = input:mean()
   local std = input:std()
   p = 0.2
   -- test forward
   local module = nn.AlphaDropout(p)
   local output = module:forward(input)
   tester:assertlt(math.abs(output:mean() - mean), 0.05, 'output mean')
   tester:assertlt(math.abs(output:std() - std), 0.05, 'output std')
   -- test backward
end

function mytest.AlphaDropout_5()
   local input = torch.randn(5000)
   local mean = input:mean()
   local std = input:std()
   p = 0.5
   -- test forward
   local module = nn.AlphaDropout(p)
   local output = module:forward(input)
   tester:assertlt(math.abs(output:mean() - mean), 0.05, 'output mean')
   tester:assertlt(math.abs(output:std() - std), 0.05, 'output std')
   -- test backward
end

function mytest.AlphaDropout_8()
   local input = torch.randn(5000)
   local mean = input:mean()
   local std = input:std()
   p = 0.8
   -- test forward
   local module = nn.AlphaDropout(p)
   local output = module:forward(input)
   tester:assertlt(math.abs(output:mean() - mean), 0.05, 'output mean')
   tester:assertlt(math.abs(output:std() - std), 0.05, 'output std')
   -- test backward
end

tester:add(mytest)
tester:run()
```

**Comment by huihuifan** [comment]
to test the back prop, you can use nn.Jacobian:

https://github.com/torch/nn/blob/master/doc/testing.md

**Comment by Naruto-Sasuke** [comment]
@kyuusaku 
```lua
function SELU(inplace)
    alpha = 1.6732632423543772848170429916717
    scale = 1.0507009873554804934193349852946
    return nn.Sequential()
        :add(nn.ELU(alpha, inplace))
        :add(nn.MulConstant(scale, true))
end

return SELU
```
so `inplace` should be true or false?
